In [1]:
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
%load_ext sql

### Query the AWS database for raw data table using 'ipython-sql'

In [2]:
def get_top_3_item_from_db(top10only):
    aItems0 = str(top10only.ITEM.values[0])
    aItems1 = str(top10only.ITEM.values[1])
    aItems2 = str(top10only.ITEM.values[2])
    
    %sql postgresql://sharknado:sharknado@sharknado-dse.ceg3hdkdq8l0.us-east-1.rds.amazonaws.com/sharknado
    local_result = %sql select distinct item,vend,l2,l3 from products \
    where item in (:aItems0,:aItems1,:aItems2)
     
    print type(local_result)
    top10only_itemnames_df = local_result.DataFrame()
    
    return top10only_itemnames_df

## 1. What are the most popular item(s) for a chosen product/category?

### Popular items for salty-snack in 2008 to 2011

In [3]:
dir_path = "/media/mqh001/Data/DSE220_Project_Iris_dataset/IRITest/"

#find saltsnck_drug *
file1 = dir_path + 'Year8/saltsnck/saltsnck_drug_1479_1530'
file2 = dir_path + 'Year9/saltsnck/saltsnck_drug_1531_1582'
file3 = dir_path + 'Year10/saltsnck/saltsnck_drug_1583_1634'
file4 = dir_path + 'Year11/saltsnck/saltsnck_drug_1635_1686'

fileList = [file1, file2, file3, file4]
fileList_size = len(fileList)

In [4]:
top3saler4years = pd.DataFrame()

for i, aFile in zip(range(1,fileList_size+1),fileList):
    panel_dr = pd.read_fwf(aFile)
    print "Year%d - panel_dr.shape ="%(i+7), panel_dr.shape
    print "   Number of items in saltysnack = ", len(panel_dr['ITEM'].unique())
    print "   Number of stores = ", len(panel_dr['IRI_KEY'].unique())
    panel_dr['DOLLARS_AVG'] = panel_dr.groupby(['ITEM'])['DOLLARS'].sum()/\
                                panel_dr.groupby(['ITEM'])['UNITS'].sum()
    
    #Get the top 3 popular items
    items = panel_dr.groupby(['ITEM'])['UNITS'].sum()
    items_df = pd.DataFrame(items)
    items_df = items_df.sort(['UNITS'], ascending=[0])
    items_df = items_df.reset_index()
    top3saler = items_df.head(3)
    
    df = get_top_3_item_from_db(top3saler)
    df = df.groupby(['item']).agg(lambda col: '    '.join(col))[['vend', 'l2','l3']]
    df['year'] = 2007+i
    df['dollar_ave'] = 0
    df['dollar_ave'][:] = panel_dr['DOLLARS_AVG'][panel_dr['ITEM'][top3saler['ITEM']]]

    #Put top10saler to 4 years table
    top3saler4years = pd.concat([top3saler4years,df])

top3saler4years['dollar_ave'] = top3saler4years['dollar_ave'].apply(lambda x: '%.2f'%x)
print "top3saler4years.shape =", top3saler4years.shape

Year8 - panel_dr.shape = (1641963, 11)
   Number of items in saltysnack =  2695
   Number of stores =  523
3 rows affected.
<class 'sql.run.ResultSet'>
Year9 - panel_dr.shape = (1635324, 11)
   Number of items in saltysnack =  2672
   Number of stores =  504
3 rows affected.
<class 'sql.run.ResultSet'>
Year10 - panel_dr.shape = (1727963, 11)
   Number of items in saltysnack =  2529
   Number of stores =  484
3 rows affected.
<class 'sql.run.ResultSet'>
Year11 - panel_dr.shape = (1801550, 11)
   Number of items in saltysnack =  2609
   Number of stores =  463
3 rows affected.
<class 'sql.run.ResultSet'>
top3saler4years.shape = (12, 5)


In [5]:
top3saler4years.reset_index().set_index(['year'])

item   vend                      l2                l3 dollar_ave
year                                                                   
2008  37705  37000            POTATO CHIPS  PROCTER & GAMBLE       1.22
2008  37717  37000            POTATO CHIPS  PROCTER & GAMBLE       1.16
2008   6452  28400  TORTILLA/TOSTADA CHIPS       PEPSICO INC       3.46
2009   7858  28400            POTATO CHIPS       PEPSICO INC       1.80
2009   7881  28400  TORTILLA/TOSTADA CHIPS       PEPSICO INC       0.98
2009   8321  28400            POTATO CHIPS       PEPSICO INC       0.98
2010   7858  28400            POTATO CHIPS       PEPSICO INC       1.79
2010   7881  28400  TORTILLA/TOSTADA CHIPS       PEPSICO INC       2.75
2010   8321  28400            POTATO CHIPS       PEPSICO INC       0.98
2011   7858  28400            POTATO CHIPS       PEPSICO INC       1.27
2011   7874  28400           CHEESE SNACKS       PEPSICO INC       1.01
2011   7881  28400  TORTILLA/TOSTADA CHIPS       PEPSICO INC       1.00

### Analysis: 
- POTATO CHIPS has been the most popular salty-snacks since 2009 to 2011.  And it also comes from one parent company 'PEPSICO INC'.

## 2. How does the popularity of an item change with the cost associated with it for a chosen year?

Answer: 
- Item 7858-POTATO CHIPS price has been cheaper over the years. 
- Item 7881-TORTILLA/TOSTADA CHIPS was cheap in 2009 than got 300% expensive on 2010, then got back about the same price as 2009 in 2011. This kinds of interesting but further investigation is not considered at this time.
- During these 4 years, Cheese Snacks only has been made in the top 3 list once in 2011. Because the IRI only limits to 2011, it could have been made beyond 2011.
- Item 6452-TORTILLA/TOSTADA CHIPS was made in popular list in 2008, however it was the most expensive item in salty snack at all time, that could be the reason why it could not make it the top 3 list in subsequent years.


## 3. How is the popularity affected over the years? 
Answers: 
- The popular items usual will be popular at least 2 years in sequence.  I also did the same analysis for 2001-2007, the top 3 popular list has changed in about 3 different types of categories over 7 years.  In 2008-2011, the top 3 popular list has maximum of 3 categories but two of them seems to repeatedly making the list.
- Price seems to play a major role.  If the price is much higher than the average of other products of the same categories, it will be dropped out the popular list.